### ISSUE TITLE
wait until async post save hook finished

### ISSUE THREAD

**Issue by andrenarchy** [Automattic/mongoose#3919]
I have a post hook that sends documents to elasticsearch after they have been saved to mongodb. However, the `doc.save()` method returns before the (async) post save hook finished its work. This results in randomly failing unit tests here because a search sometimes is run before the documents have been indexed by elasticsearch. Note: this is not an issue with elasticsearch (I'm calling [refresh](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference-2-2.html#api-indices-refresh-2-2) before executing a search).

Example with timeout instead of elasticsearch:

```
schema = new mongoose.Schema({title: String});
schema.post('save', function(doc, next) {
  setTimeout(next, 2000);
});
mongoose.model('Book', schema);
mongoose.model('Book').create({title: 'test'}, function (err, doc) {
  console.log('done');
});
```

The above code immediately prints `done`. The expected behavior is that it waits 2s and then prints `done`.

**Comment by nschloe** [comment]
Definitely an annoyance for me too. Right now, I'm throwing manual timeouts all over the place; needless to say that this is not contributing to a nice development workflow.

**Comment by basiclaser** [comment]
+1 !!!

**Comment by bitgal** [comment]
here too :/

**Comment by andrenarchy** [comment]
Any news here? This issue is really annoying because it makes our unit tests fail randomly.

**Comment by vkarpov15** [comment]
Not yet. This is a pretty significant change that will require a hefty bit of work because of how post hooks are structured internally.

**Comment by ohjames** [comment]
At the very least is it possible to provide some way to wait on all the outstanding "post" hooks to finish, if I could wait on that after each save it would allow my tests to work in a sane way.

**Comment by vkarpov15** [comment]
@ohjames no good way to do that atm, you can hack it by adding a post hook that executes last that resolves a promise but that's not ideal

**Comment by pedrovagner** [comment]
Why I can't reproduce this issue? I'm using `mongoose 4.6.0` and `doc.save()` is executed only after async post finishes (BTW, this is my desired behavior). I end up here, because I want make sure that `doc.save()` is executed only after async post finishes. Now I'm confusing. Just note that in my case, I'm using `doc.save()`, not `Model.create()`.

**Comment by vkarpov15** [comment]
@pedrovagner can you please clarify your question? I don't understand

**Comment by pedrovagner** [comment]
When I execute:

```javascript
schema = new mongoose.Schema({title: String});
schema.post('save', function(doc, next) {
  setTimeout(next, 2000);
});
mongoose.model('Book', schema);
mongoose.model('Book').create({title: 'test'}, function (err, doc) {
  console.log('done');
});
```
`done` isn't printed immediately, it's printed only after 2s. This behavior is different as described by @andrenarchy. Why? I'm confusing.

**Comment by vkarpov15** [comment]
That's curious, I'm not able to reproduce this issue in any 4.x version. @andrenarchy do you remember what version this issue was for?

**Comment by andrenarchy** [comment]
@vkarpov15 no, I'm not sure which version this was.

But actually my original example doesn't print anything at all with `mongoose@4.7.2`. Is the callback even called?

**Comment by vkarpov15** [comment]
Hmm works for me:

```javascript
'use strict';

var assert = require('assert');
var mongoose = require('mongoose');
var Schema = mongoose.Schema;

mongoose.connect('mongodb://localhost/gh3919');
mongoose.set('debug', true);

var schema = new mongoose.Schema({title: String});
schema.post('save', function(doc, next) {
  setTimeout(() => { console.log('hi'); next(); }, 2000);
});
mongoose.model('Book', schema);
mongoose.model('Book').create({title: 'test'}, function (err, doc) {
  console.log('done');
});
```

```
$ rm -rf node_modules/mongoose && npm install mongoose@4.7.2
npm WARN deprecated mongodb@2.2.12: Please use 2.2.16 or higher due to a regression in updateOne/Many upsertedId return value
/home/val/Workspace/10gen/troubleshoot-mongoose
└─┬ mongoose@4.7.2 
  ├── async@2.1.4 
  ├── bson@0.5.7 
  ├── hooks-fixed@1.2.0 
  ├── kareem@1.1.3 
  ├─┬ mongodb@2.2.12 
  │ ├── es6-promise@3.2.1 
  │ ├── mongodb-core@2.0.14 
  │ └─┬ readable-stream@2.1.5 
  │   └── isarray@1.0.0 
  ├── mpath@0.2.1 
  ├── mpromise@0.5.5 
  ├─┬ mquery@2.0.0 
  │ ├── bluebird@2.10.2 
  │ └── sliced@0.0.5 
  ├── ms@0.7.2 
  ├── muri@1.1.1 
  └── sliced@1.0.1 

npm WARN enoent ENOENT: no such file or directory, open '/home/val/Workspace/10gen/troubleshoot-mongoose/package.json'
npm WARN troubleshoot-mongoose No description
npm WARN troubleshoot-mongoose No repository field.
npm WARN troubleshoot-mongoose No README data
npm WARN troubleshoot-mongoose No license field.
$ node gh-3919.js 
Mongoose: books.insert({ title: 'test', _id: ObjectId("5854b78bc1ddd21907e03b4d"), __v: 0 })
hi
done
^C
$ 
```

Do you get the debug message when you run the script on your machine?

**Comment by andrenarchy** [comment]
OK, now it works... seems to have been a problem with the mongo connection.
